### ISSUE TITLE
[BUG] Android "hardware" back button minimizes/closes the app

### ISSUE THREAD

**Issue by VNGames** [PrismLibrary/Prism#2990]
### Description

Created a simple MAUI app (NET 7.0). Added Prism (8.1.273-pre).
Starting app with `NavigationPage/MainPage`
Then Pushing `Page1`
Then Pushing `Page2`
Tapping on back button closes both pages and closes/minimizes the app.
UPD: If I add Page3 on a stack it closes Page 3 an Page 2 and minimizes the app. 
Definitely double back button handling:
I think one is done by MAUI, second one by Prism. Maybe this is related https://github.com/dotnet/maui/issues/15296#issuecomment-1586154483

Plus something closes the app afterwards (Probably Prism because with the same config of pages using only maui for navigation no problems encountered (unless I add `android:enableOnBackInvokedCallback="true" into manifest, then no navigation happens at all.)).

If I add this into manifest while using Prism
`android:enableOnBackInvokedCallback="true"`
When tapping on back button it doesn't close any page but closes/minimizes the app (Again pointing to Prism doing that)

Here is a project to reproduce this
[PrismBackTest.zip](https://github.com/PrismLibrary/Prism/files/13177569/PrismBackTest.zip)

### Steps to Reproduce

1. Build and launch the app.
2. Tap on `Click me` button
3. Tap on `Navigate to Page 2` button
4. Tap on hardware back button

### Platform with bug

.NET MAUI

### Affected platforms

Android

### Did you find any workaround?

No workaround.

**Comment by maonaoda** [comment]
This is a serious hindrance

**Comment by maonaoda** [comment]
@dansiegel 
`application.Quit(); `is the reason.

---
why should call application.Quit()? instead of going back to MainPage from Page1 according ↓
https://github.com/PrismLibrary/Prism/blob/4167e97a56ed3e6fe996341b49eed021911d2678/src/Maui/Prism.Maui/PrismAppBuilder.cs#L66C1-L66C1
```
                android.OnBackPressed(activity =>
                {
                    var root = ContainerLocator.Container;
                    if (root is null)
                        return false;

                    var app = root.Resolve<IApplication>();
                    var windows = app.Windows.OfType<PrismWindow>();
                    if (!windows.Any(x => x.IsActive))
                        return false;

                    var window = windows.First(x => x.IsActive);
                    if(window.IsRootPage && app is Application application)
                    {
                        application.Quit();
                        return false;
                    }

                    window.OnSystemBack();

                    return true;
                });
```
---
and  Quit app instead of going background from MainPage by clicking "hardware" back button according ↓
https://github.com/PrismLibrary/Prism/blob/4167e97a56ed3e6fe996341b49eed021911d2678/src/Maui/Prism.Maui/Common/MvvmHelpers.cs#L306

**Comment by maonaoda** [comment]
What I expect is this:

https://github.com/PrismLibrary/Prism/assets/32831595/1ffca96e-b8e2-4ba2-a62c-7b359c901206

and need 

- remove  application.Quit();
https://github.com/PrismLibrary/Prism/blob/4167e97a56ed3e6fe996341b49eed021911d2678/src/Maui/Prism.Maui/PrismAppBuilder.cs#L66C44-L66C44

- change this method 
https://github.com/PrismLibrary/Prism/blob/4167e97a56ed3e6fe996341b49eed021911d2678/src/Maui/Prism.Maui/Controls/PrismNavigationPage.cs#L32-L36

to
```
    protected override bool OnBackButtonPressed()
    {
        if (CurrentPage.SendBackButtonPressed())
            return true;

        if (NavigationPageController.StackDepth > 1)
        {
            BackButtonPressed.Invoke(this, EventArgs.Empty);
            return true;
        }

        return false;
    }
```

※my be 
https://github.com/PrismLibrary/Prism/blob/4167e97a56ed3e6fe996341b49eed021911d2678/src/Maui/Prism.Maui/Common/MvvmHelpers.cs#L306

**Comment by maonaoda** [comment]
> But when NavigationStack is NavigationPage/TabbedPage,It will not be able to go background,thats needs further fix.
https://github.com/PrismLibrary/Prism/blob/4167e97a56ed3e6fe996341b49eed021911d2678/src/Maui/Prism.Maui/PrismAppBuilder.cs#L50-L73

window.IsRootPage need change 
from 
```
    internal bool IsRootPage => Page switch
    {
        TabbedPage tabbed => tabbed.CurrentPage,
        NavigationPage nav => nav.RootPage,
        _ => Page
    } == CurrentPage;
```
to
```
    internal bool IsRootPage => GetCurrentPage(Page) == CurrentPage;

    private Page GetCurrentPage(Page page)
    {
        return page switch
        {
            TabbedPage tabbed => tabbed.CurrentPage,
            NavigationPage nav => nav.RootPage is TabbedPage tabbedPage ? GetCurrentPage(tabbedPage.CurrentPage) : nav.RootPage,
            _ => page
        };
    }
```

---
@dansiegel 
For now I can only use this local change temporarily and look forward to further modifications from Prism.

**Comment by VNGames** [comment]
Net 8.0 came out. And when using it in combination with Prism.DryIoc.Maui 9.0.271-pre it works fine.

**Comment by denizz-ryzhkov** [comment]
@VNGames Great news. Gonna try it out

**Comment by VNGames** [comment]
@denizz-ryzhkov I Migrated to Net 8.0. then updated Prism. Try this project.
[PrismBackTest.zip](https://github.com/PrismLibrary/Prism/files/13367722/PrismBackTest.zip)

**Comment by thomasgalliker** [comment]
There was no code change in the affected piece of code. So, this behavior still exists.

**Comment by VNGames** [comment]
@denizz-ryzhkov you are right. It got better but still closes the app when you are on the second page.
here is the project to play if anybody wants
[PrismBackTest.zip](https://github.com/PrismLibrary/Prism/files/13367750/PrismBackTest.zip)

**Comment by dweyermann** [comment]
I am experiencing this as well with .NET 8.0 and Prism.DryIoc.Maui 9.0.271-pre.

**Comment by kyurkchyan** [comment]
I am having the same problem on my Android app. Hardware back navigation kills the app. Instead, it should move it to the background.

**Comment by dansiegel** [comment]
Thanks @maonaoda... from my testing removing application.Quit() does seem to resolve the issue... I did test out your other suggestions but the behavior was not correct at all.

**Comment by maonaoda** [comment]
@dansiegel 
thanks for yours fix. I will do test on my app.

**Comment by maonaoda** [comment]
@dansiegel 
 removing application.Quit() had fixed this bug yet.

>  I did test out your other suggestions but the behavior was not correct at all.
Is there any problem?

the further bug of going back from Tabbedpage still exists.
I will create a new issue and add a reproduction。

**Comment by Baraiboapex** [comment]
Does anyone have at least a work around for this issue?

**Comment by Baraiboapex** [comment]
Has this been fixed yet?

**Comment by LasseWolter** [comment]
@Baraiboapex That's expected. 
I don't think the fix was released as part of 9.0.401. 

Release 9.0.401 was released on 14/02/24. 
The [PR](https://github.com/PrismLibrary/Prism/pull/3085) with the fix was merged into master on 24/02/24.
<img width="688" alt="image" src="https://github.com/PrismLibrary/Prism/assets/29123172/32b96f80-47ad-41d0-8cf8-d3701105db44">

**Comment by dansiegel** [comment]
Do keep in mind that those with the Commercial Plus license already have access to this fix among others.

**Comment by Baraiboapex** [comment]
Okay. When is the next release scheduled to come out?

**Comment by dansiegel** [comment]
It will be out when it we are ready. We don't have a hard date, it's exceptionally rare that we set a hard date unless someone is paying for features that need to be available by a given date.

**Comment by tbaggett** [comment]
I've updated to 9.0.436-pre and still encounter a NavigationException with the message "Cannot Pop Application MainPage" at Prism.Navigation.PageNavigationService.GoBackInternalAsync(INavigationParameters parameters).

We have a slightly different page hierarchy than the example provided in this ticket. @dansiegel I'm happy to create a separate ticket with its own repro project if you need it.

Our page hierarchy:

NavigationPage
- TabbedPage
-- NavigationPage
--- Root ContentPage
-- NavigationPage
--- Root ContentPage
-- NavigationPage
--- Root ContentPage

The root NavigationPage placed before the TabbedPage is needed to show some pages without the tab bar being present. 

The repro steps I'm taking are:

1. Launch app
2. Switch to the last tab
3. Tap an option on the tab's root page that causes a second page to be pushed onto that tab's NavigationPage stack
4. Tap an option on the tab's second page that causes a third page to be push onto the tab's NavigationPage stack
5. Tap the Android hardware back button. Exception occurs and app closes.

**Comment by maonaoda** [comment]
@tbaggett 
may be you can use my workaround (BaseNavigationPage) to replace PrismNavigationPage 
under  9.0.436-pre,and tell me whether it works. :)

```
    public class BaseNavigationPage : PrismNavigationPage
    {
#if ANDROID
        INavigationPageController NavigationPageController => this;

        public new event EventHandler BackButtonPressed;

        public BaseNavigationPage()
        {
            BackButtonPressed += HandleBackButtonPressed;
            BarTextColor = Colors.White;
        }

        public BaseNavigationPage(Page page)
            : base(page)
        {
            BackButtonPressed += HandleBackButtonPressed;
            BarTextColor = Colors.White;
        }

        private async void HandleBackButtonPressed(object sender, EventArgs args)
        {
            await MvvmHelpers.HandleNavigationPageGoBack(this);
        }
#else
        public BaseNavigationPage()
        {
            BarTextColor = Colors.White;
        }

        public BaseNavigationPage(Page page)
            : base(page)
        {
            BarTextColor = Colors.White;
        }

#endif

        protected override bool OnBackButtonPressed()
        {
#if ANDROID
            if (CurrentPage.SendBackButtonPressed())
                return true;

            if (NavigationPageController.StackDepth > 1)
            {
                BackButtonPressed.Invoke(this, EventArgs.Empty);
                return true;
            }

            return false;
#else
            return base.OnBackButtonPressed();
#endif
        }
    }
```

**Comment by tbaggett** [comment]
> @tbaggett may be you can use my workaround (BaseNavigationPage) to replace PrismNavigationPage under 9.0.436-pre,and tell me whether it works. :)
> 
> like https://github.com/maonaoda/Prism/tree/bugfix/1/android-back-fix-confirm/tests/PrismBugs
> 
> ```
>     public class BaseNavigationPage : PrismNavigationPage
>     {
> #if ANDROID
>         INavigationPageController NavigationPageController => this;
> 
>         public new event EventHandler BackButtonPressed;
> 
>         public BaseNavigationPage()
>         {
>             BackButtonPressed += HandleBackButtonPressed;
>         }
> 
>         public BaseNavigationPage(Page page)
>             : base(page)
>         {
>             BackButtonPressed += HandleBackButtonPressed;
>         }
> 
>         private async void HandleBackButtonPressed(object sender, EventArgs args)
>         {
>             await MvvmHelpers.HandleNavigationPageGoBack(this);
>         }
> #else
>         public BaseNavigationPage()
>         {
>         }
> 
>         public BaseNavigationPage(Page page)
>             : base(page)
>         {
>         }
> 
> #endif
> 
>         protected override bool OnBackButtonPressed()
>         {
> #if ANDROID
>             if (CurrentPage.SendBackButtonPressed())
>                 return true;
> 
>             if (NavigationPageController.StackDepth > 1)
>             {
>                 BackButtonPressed.Invoke(this, EventArgs.Empty);
>                 return true;
>             }
> 
>             return false;
> #else
>             return base.OnBackButtonPressed();
> #endif
>         }
>     }
> ```

Thanks so much for sharing that, @maonaoda ! I added it and yes, it does resolve my issue! Much appreciated.

**Comment by Baraiboapex** [comment]
Where would I find the private discord server after a commercial license has been purchased?

**Comment by Baraiboapex** [comment]
> > @tbaggett may be you can use my workaround (BaseNavigationPage) to replace PrismNavigationPage under 9.0.436-pre,and tell me whether it works. :)
> > like https://github.com/maonaoda/Prism/tree/bugfix/1/android-back-fix-confirm/tests/PrismBugs
> > ```
> >     public class BaseNavigationPage : PrismNavigationPage
> >     {
> > #if ANDROID
> >         INavigationPageController NavigationPageController => this;
> > 
> >         public new event EventHandler BackButtonPressed;
> > 
> >         public BaseNavigationPage()
> >         {
> >             BackButtonPressed += HandleBackButtonPressed;
> >         }
> > 
> >         public BaseNavigationPage(Page page)
> >             : base(page)
> >         {
> >             BackButtonPressed += HandleBackButtonPressed;
> >         }
> > 
> >         private async void HandleBackButtonPressed(object sender, EventArgs args)
> >         {
> >             await MvvmHelpers.HandleNavigationPageGoBack(this);
> >         }
> > #else
> >         public BaseNavigationPage()
> >         {
> >         }
> > 
> >         public BaseNavigationPage(Page page)
> >             : base(page)
> >         {
> >         }
> > 
> > #endif
> > 
> >         protected override bool OnBackButtonPressed()
> >         {
> > #if ANDROID
> >             if (CurrentPage.SendBackButtonPressed())
> >                 return true;
> > 
> >             if (NavigationPageController.StackDepth > 1)
> >             {
> >                 BackButtonPressed.Invoke(this, EventArgs.Empty);
> >                 return true;
> >             }
> > 
> >             return false;
> > #else
> >             return base.OnBackButtonPressed();
> > #endif
> >         }
> >     }
> > ```
> 
> Thanks so much for sharing that, @maonaoda ! I added it and yes, it does resolve my issue! Much appreciated.

Also, is this just plug and play for a current prism project or is there extra overhead that I need to worry about?

**Comment by tbaggett** [comment]
> Where would I find the private discord server after a commercial license has been purchased?

The email used to purchase the license will receive an invitation to the Discord server

**Comment by dansiegel** [comment]
> Where would I find the private discord server after a commercial license has been purchased?

With the new system... when you buy the license or licenses you will need to assign the license (even if it's to yourself)... if you have a license assigned to your account then you should see a link to add your discord username and then join the discord server

**Comment by maonaoda** [comment]
```
Also, @maonaoda, is this just plug and play for a current prism project or is there extra overhead that I need to worry about?
```

This is only a fix for Android`s OnBackButtonPressed. I'm not sure what extra consumption you're talking about.
-  the shutdown of the app on Android
-  SendBackButtonPressed on Android

**Comment by Baraiboapex** [comment]
@maonaoda, Thanks. I guess my question is where you would put this code in order for this fix to work. I only asked about extra overhead because I was just wondering if this class needed some extra pieces of code. I apologize for the lack of clarity.

**Comment by Baraiboapex** [comment]
@maonaoda I was asking essentially if this code was simply just something you drop in or if there is some extra work needed to get it to work. I guess my next question is where would you put this code?

**Comment by Baraiboapex** [comment]
One last question, @moandoa we can use this fix in a commercial product no strings attached, correct? I mean, it looks like this is obviously the case, but just want to be sure ;)

**Comment by maonaoda** [comment]
> One last question, @maonaoda we can use this fix from your github link in a commercial product (Just the BaseNavigationPage class to be sepcific) no strings attached, correct? I mean, it looks like this is obviously the case, but just want to be sure ;)
> 

Yes, of course, that's the override

**Comment by Baraiboapex** [comment]
@maonaoda  Thanks. I figured I didn't see a license or anything I can just be a stickler for this stuff. LOL

**Comment by maonaoda** [comment]
This [PR 3148 ](https://github.com/PrismLibrary/Prism/pull/3148)causes the app to no longer be able to return to the background using https://github.com/PrismLibrary/Prism/issues/2990#issuecomment-2027892817

And this PR does not seem to fix the current problem。：）

**Comment by bodeg** [comment]
> Is there any workaround for Community license users with 9.0.401-pre?

Maybe you can try something like the following on your BaseContentPage:

```
protected override bool OnBackButtonPressed()
{
  if (ViewModel != null)
  {
    // Use a timer to delay the execution of the command
    Application.Current?.Dispatcher.StartTimer(TimeSpan.FromMilliseconds(100), () =>
    {
      MainThread.BeginInvokeOnMainThread(() => { ViewModel.BackNavigationCommand.Execute(); });
      // Return false to run the timer callback only once
      return false;
    });
    // Cancel back navigation
    return true;
  }
  // Cancel back navigation
  return true;
}

```
